assertj-error-prone/src/main/java/com/palantir/assertj/errorprone/AssertjAssertThatThrownBy.java

```diff
@@ -0,0 +1,118 @@
+        name = "AssertjAssertThatThrownBy",
+        link = "https://github.com/palantir/assertj-automation",
+        linkType = BugPattern.LinkType.CUSTOM,
+        providesFix = BugPattern.ProvidesFix.REQUIRES_HUMAN_ATTENTION,
+        severity = BugPattern.SeverityLevel.SUGGESTION,
+        summary = "Prefer AssertJ assertThatThrownBy assertions over try-catch with fail statements")
+public final class AssertjAssertThatThrownBy extends BugChecker implements BugChecker.TryTreeMatcher {
+    private static final ImmutableList<String> IGNORED_FAIL_MESSAGES = ImmutableList.of("\"fail\"", "\"\"");
+
+    private static final Matcher<StatementTree> FAIL_METHOD = Matchers.anyOf(
+            Matchers.expressionStatement(
+                    MethodMatchers.staticMethod().onClass("org.junit.Assert").named("fail")),
+            Matchers.expressionStatement(MethodMatchers.staticMethod()
+                    .onClass("org.junit.jupiter.api.Assertions")
+                    .named("fail")),
+            Matchers.expressionStatement(MethodMatchers.staticMethod()
+                    .onClass("org.assertj.core.api.Assertions")
+                    .named("fail")));
+
+    @Override
+    public Description matchTry(TryTree tree, VisitorState state) {
+        List<? extends StatementTree> tryStatements = tree.getBlock().getStatements();
+        if (tryStatements.isEmpty() || tree.getCatches().size() != 1 || tree.getFinallyBlock() != null) {
+            return Description.NO_MATCH;
+        }
+        CatchTree catchTree = Iterables.getOnlyElement(tree.getCatches());
+        if (!catchTree.getBlock().getStatements().isEmpty()
+                || catchTree.getParameter().getType().getKind() == Tree.Kind.UNION_TYPE) {
+            return Description.NO_MATCH;
+        }
+        StatementTree lastStatement = Iterables.getLast(tryStatements);
+        List<? extends StatementTree> throwingStatements = tryStatements.subList(0, tryStatements.size() - 1);
+        if (!FAIL_METHOD.matches(lastStatement, state) || throwingStatements.isEmpty()) {
+            return Description.NO_MATCH;
+        }
+        if (!TestCheckUtils.isTestCode(state)) {
+            return Description.NO_MATCH;
+        }
+        Optional<String> failMessage = getFailMessage(lastStatement, state);
+        return buildDescription(tree)
+                .addFix(tryFailToAssertThatThrownBy(
+                        tree, throwingStatements, catchTree.getParameter(), failMessage, state))
+                .build();
+    }
+
+    private static Optional<String> getFailMessage(StatementTree failStatement, VisitorState state) {
+        Iterable<? extends ExpressionTree> failArgs =
+                ((MethodInvocationTree) ((ExpressionStatementTree) failStatement).getExpression()).getArguments();
+        return Optional.ofNullable(Iterables.get(failArgs, 0, null))
+                .map(state::getSourceForNode)
+                .filter(msg -> !IGNORED_FAIL_MESSAGES.contains(msg));
+    }
```

Comment:
We could probably simplify this and ignore the fail message entirely. In most cases it just tells us the same thing as the dynamic assertj failure. Thoughts?

Reply:
I dunno, I kinda prefer for our refactors to preserve user-authored stuff where possible, as i think we lose trust in the very rare cases where we delete something that was actually important... e.g. 

```java
       try {
            foo.getPhoneNumber();
            fail("Getting the phone number of an interface without an annotation should fail");
        } catch (IllegalArgumentException e) {

        }
```

Reply:
That’s a good point :+1: